tempesta_fw/ss_skb.c

```diff
@@ -164,129 +162,155 @@ __lookup_pgfrag_room(struct sk_buff *skb, int len)
 
 /**
  * Somewhat like skb_shift().
+ * Make room for @n fragments starting with slot @from.
  *
- * Beware: @from can be equal to MAX_SKB_FRAGS if we need to insert a new
- * fragment after the last one.
+ * Beware: @from can be equal to MAX_SKB_FRAGS when a new fragment
+ * is inserted after the last one.
+ *
+ * @return 0 on success, -errno on failure.
+ * @return New SKB in @it->skb if new SKB is allocated.
  */
 static int
-__extend_pgfrags(struct sk_buff *skb, struct sk_buff *pskb, int from, int n)
+__extend_pgfrags(struct sk_buff *skb, int from, int n, TfwStr *it)
 {
-	int i, n_frag = 0;
-	struct skb_shared_info *psi, *si = skb_shinfo(skb);
+	int i, n_shift, n_excess = 0;
+	struct skb_shared_info *si = skb_shinfo(skb);
 
-	if (skb_shinfo(skb)->nr_frags > MAX_SKB_FRAGS - n) {
+	BUG_ON(from > si->nr_frags);
+
+	/* No room for @n extra fragments in the SKB. */
+	if (si->nr_frags + n > MAX_SKB_FRAGS) {
 		skb_frag_t *f;
-		struct sk_buff *skb_frag;
+		struct sk_buff *nskb;
 
-		psi = pskb ? skb_shinfo(pskb) : si;
-		skb_frag = psi->frag_list;
-		n_frag = skb_shinfo(skb)->nr_frags + n - MAX_SKB_FRAGS;
+		BUG_ON(skb_has_frag_list(skb));
 
-		if (skb_frag && !skb_headlen(skb_frag)
-		    && skb_shinfo(skb_frag)->nr_frags <= MAX_SKB_FRAGS - n_frag)
-		{
-			int r = __extend_pgfrags(skb_frag, NULL, 0, n_frag);
-			if (r)
-				return r;
-		} else {
-			skb_frag = alloc_skb(0, GFP_ATOMIC);
-			if (!skb_frag)
-				return -ENOMEM;
-			skb_frag->next = psi->frag_list;
-			psi->frag_list = skb_frag;
-		}
+		/*
+		 * Allocate a new SKB to hold @n_excess fragments.
+		 * Put it on frag_list of the current SKB.
+		 */
+		nskb = alloc_skb(MAX_TCP_HEADER, GFP_ATOMIC);
+		if (nskb == NULL)
+			return -ENOMEM;
+		skb_reserve(nskb, MAX_TCP_HEADER);
```

Comment:
It turns out that when these SKBs get into TCP/IP stack for sending, the stack expects that SKBs have sufficient room to do an `skb_pull()`. If there's no headroom, it crashes in skb_pull().

The total absence of headroom is okay for SKB fragments (from `frag_list`), but it is not in this case.